File: my-folder/0042-trapping-rain-water/solution.py
class Solution:
    def trap(self, height: List[int]) -> int:
        len_height = len(height) 
        pre = [0] * (len_height)
        suff = [0] * (len_height)
        pre[0] = height[0]
        
        for i in range(1, len_height):
            pre[i] = max(pre[i - 1], height[i])
        
        suff[-1] = height[-1]
        for i in range(len_height - 2, -1 , -1):
            suff[i] = max(height[i], suff[i + 1])
        
        water_trapped = 0
        for i in range(len_height):
            water_trapped += abs(min(suff[i], pre[i]) - height[i])

        return water_trapped
            

        

File: my-folder/0042-trapping-rain-water/solution.java
class Solution {
    public int trap(int[] height) {
        int lmax = 0, rmax =  0, waterTrapped = 0;
        int left = 0;
        int right = height.length - 1;

        while(left < right){
            if(height[left] < height[right]){
                if(height[left] >= lmax){
                    lmax = height[left];
                }
                else{
                    waterTrapped += lmax - height[left];
                }
                left++;
            }
            else{
                if(height[right] >= rmax){
                    rmax = height[right];
                }
                else{
                    waterTrapped += rmax - height[right];
                }
                right--;
            }
        }

        return waterTrapped;
    }
}

//                      R
// [0,1,0,2,1,0,1,3,2,1,2,1]
//    L

// [0 1 1 2 2 2 2 3 3 3 3 3] - PRE
// [3,3,3,3,3,3,3,3,2,2,2,1] - SUFF
// [0 1 1 2 2 2 2 0 2 2 2 1] - SUB
